README.md
# flutter_appears_randomly

Animation with random gradients

随机渐变出现的动画

![](https://github.com/Wing-Li/flutter_appears_randomly/blob/master/res/example_video.gif)

## 使用示例

    Container(
      padding: EdgeInsets.symmetric(horizontal: 16), // 为了防止贴边不美观。（可以去掉）
      child: AppearsRandomlyWidget(
        width: screenWidth, // 整体的宽高。用于计算 item 出现的范围
        height: screenWidth,
        
        itemWidth: 56, // 每个 item 的宽高。用于计算 item 的出现
        itemHeight: 56,
        itemSizeRange: 16, // 从小变大时，最大/最小的缩放范围。（会生成随机值，此处为最大值）
        
        centerRestrictedArea: Size(iconWidth, iconWidth),  // 中心布局的范围
        centerChild: Container( // 中心布局。（可为空）
          width: iconWidth,
          height: iconWidth,
          alignment: Alignment.center,
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(120),
            border: Border.all(color: Colors.white, width: 3),
          ),
          child: ClipOval(
            child: Image.asset(_getImage("header", format: "webp"), width: iconWidth, height: iconWidth),
          ),
        ),
        
        itemCount: list.length, // 总数
        itemBuilder: (context, index) { // item 布局
          var imgRes = list[index];
          return GestureDetector(
            child: Container(
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(56),
                border: Border.all(color: Colors.white, width: 0.5),
              ),
              child: ClipOval(child: Image.asset(_getImage(imgRes), width: 52, height: 52)),
            ),
            onTap: () {
              // TODO: 点击
            },
          );
        },
      ),
    ),




CHANGELOG.md
## 0.0.2

* Initial first release

lib/stagger_animation.dart
import 'dart:math';

import 'package:flutter/material.dart';

class StaggerAnimation extends StatefulWidget {
  final Widget child;
  final double width;
  final double height;
  final double sizeRange;
  final int animMilliseconds;
  final Function onEnd;

  const StaggerAnimation({
    Key key,
    this.width,
    this.height,
    this.sizeRange,
    this.child,
    this.onEnd,
    this.animMilliseconds = 4000,
  }) : super(key: key);

  @override
  _StaggerAnimationState createState() => _StaggerAnimationState();
}

class _StaggerAnimationState extends State<StaggerAnimation> with SingleTickerProviderStateMixin {
  AnimationController _animationController;
  Animation<Size> _sizeTween;
  Animation<double> _opacity;

  bool isEnd = false;

  @override
  void initState() {
    super.initState();

    isEnd = false;

    _animationController = AnimationController(duration: Duration(milliseconds: widget.animMilliseconds), vsync: this)
      ..addListener(() {
        if (_animationController.status == AnimationStatus.dismissed) {
          setState(() {
            isEnd = true;
          });
          if (widget.onEnd != null) widget.onEnd();
        }
      });

    var random = widget.sizeRange * Random().nextDouble();
    _sizeTween = SizeTween(
      begin: Size(widget.width - random, widget.height - random),
      end: Size(widget.width + random, widget.height + random),
    ).animate(_animationController);

    _opacity = Tween<double>(begin: 0.0, end: 0.9).animate(
      // _animationController
      CurvedAnimation(
        parent: _animationController,
        curve: Interval(
          0.0, 0.4, //间隔，前60%的动画时间
          curve: Curves.ease,
        ),
      ),
    );

    _playAnimation();
  }

  Future<Null> _playAnimation() async {
    try {
      //先正向执行动画
      await _animationController.forward().orCancel;
      //再反向执行动画
      await _animationController.reverse().orCancel;
    } on TickerCanceled {
      // the animation got canceled, probably because we were disposed
    }
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return isEnd
        ? SizedBox()
        : AnimatedBuilder(
            animation: _animationController,
            builder: (BuildContext context, Widget c) {
              return Opacity(
                opacity: _opacity.value,
                child: SizedBox(
                  width: _sizeTween.value.width,
                  height: _sizeTween.value.height,
                  child: widget.child,
                ),
              );
            },
          );
  }
}


lib/appears_randomly_widget.dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';

import 'stagger_animation.dart';

class AppearsRandomlyWidget extends StatefulWidget {
  final double width;
  final double height;
  final int itemCount;
  final ItemBuilder itemBuilder;
  final double itemWidth;
  final double itemHeight;
  final double itemSizeRange;
  final Size centerRestrictedArea;
  final Widget centerChild;

  /// 间隔时间，毫秒
  final int intervals;

  /// 动画时间，毫秒
  final int animMilliseconds;

  const AppearsRandomlyWidget({
    Key key,
    this.itemCount,
    this.itemBuilder,
    this.width,
    this.height,
    this.itemWidth,
    this.itemHeight,
    this.itemSizeRange,
    this.intervals = 2000,
    this.animMilliseconds = 5000,
    this.centerRestrictedArea,
    this.centerChild,
  }) : super(key: key);

  @override
  _AppearsRandomlyWidgetState createState() => _AppearsRandomlyWidgetState();
}

class _AppearsRandomlyWidgetState extends State<AppearsRandomlyWidget> {
  List<Widget> childList = [];
  BuildContext mContext;

  /// 控件的key
  GlobalKey _bodyKey = GlobalKey();
  double width;
  double height;

  List<Size> savedPoint = [];

  @override
  void initState() {
    super.initState();

    _initState();

    Future.delayed(Duration.zero, _loopMain);
  }

  _initState() {
    width = widget.width;
    height = widget.height;

    // 在控件渲染完成后执行的回调
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _findRenderObject();
    });

    if (widget.centerChild != null) {
      if (childList.length > 0) {
        childList.removeAt(0);
      }
      childList.insert(
          0,
          Positioned(
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            child: Center(child: widget.centerChild),
          ));
    }
  }

  /// 得到当前控件的宽高
  _findRenderObject() {
    RenderBox renderBox = _bodyKey.currentContext.findRenderObject();
    setState(() {
      width = renderBox.size.width;
      height = renderBox.size.height;
    });
  }

  /// 获取当前显示的图标位置
  Size _getShowXY() {
    double x = (width - widget.itemWidth) * Random().nextDouble() * 0.95;
    double y = (height - widget.itemHeight) * Random().nextDouble() * 0.95;

    try {
      if (widget.centerRestrictedArea != null) {
        double areaW = widget.centerRestrictedArea.width / 2 + widget.itemWidth / 2;
        double areaH = widget.centerRestrictedArea.height / 2 + widget.itemHeight / 2;
        // 随机值不能在中心禁区
        if (width / 2 - areaW - widget.itemWidth * 0.5 - widget.itemSizeRange < x &&
            x < width / 2 + areaW &&
            height / 2 - areaH - widget.itemHeight * 0.5 - widget.itemSizeRange < y &&
            y < height / 2 + areaH) {
          return _getShowXY();
        }
      }

      // 随机值，不能在已有的列表里
      for (int i = 0; i < savedPoint.length; i++) {
        var it = savedPoint[i];
        if (it.width - widget.itemWidth - widget.itemSizeRange < x &&
            x < it.width + widget.itemWidth + widget.itemSizeRange && //
            it.height - widget.itemHeight - widget.itemSizeRange < y &&
            y < it.height + widget.itemHeight + widget.itemSizeRange) {
          return Size(-100, -100);
        }
      }

      return Size(x, y);
    } catch (e) {
      // MyUtils.log("$x - $y == ${savedPoint.length}");

      return Size(x, y);
    }
  }

  _loopMain({int delayed = 30}) {
    Future.delayed(Duration(milliseconds: delayed), () {
      if (mContext != null) {
        if (widget.itemCount > 0) {
          int count = _random(1, 4);
          for (int i = 0; i < count; i++) {
            Size saveXY = _getShowXY();
            if (saveXY.width < 0) continue;

            savedPoint.add(saveXY);

            var childItem = Positioned(
              left: saveXY.width,
              bottom: saveXY.height,
              child: _buildChildItem(
                context,
                onEnd: () {
                  // MyUtils.log("end");
                  savedPoint.remove(saveXY);
                },
              ),
            );
            childList.add(childItem);
          }

          if (!isDispose) setState(() {});
        }
      }

      int delayed = _random(300, widget.intervals);
      _loopMain(delayed: delayed);
    });
  }

  _buildChildItem(BuildContext context, {Function onEnd}) {
    int randomIndex = Random().nextInt(widget.itemCount);

    return StaggerAnimation(
      width: widget.itemWidth,
      height: widget.itemHeight,
      sizeRange: widget.itemSizeRange,
      child: widget.itemBuilder(context, randomIndex),
      animMilliseconds: widget.animMilliseconds,
      onEnd: onEnd,
    );
  }

  @override
  void didUpdateWidget(covariant AppearsRandomlyWidget oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.width != oldWidget.width || widget.height != oldWidget.height) {
      _initState();
    }
  }

  bool isDispose = false;

  @override
  void dispose() {
    isDispose = true;
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    mContext = context;

    return Container(
      key: _bodyKey,
      width: width,
      height: height,
      child: Stack(
        children: childList,
      ),
    );
  }

  int _random(int start, int end) {
    Random random = new Random();
    return random.nextInt(end) % (end - start + 1) + start;
  }
}

typedef ItemBuilder = Widget Function(BuildContext context, int index);


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_appears_randomly/appears_randomly_widget.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({
    Key key,
  }) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double screenWidth = 375;

  List<String> list = [];

  @override
  void initState() {
    super.initState();

    list = List.generate(30, (index) => "img$index");
  }

  @override
  Widget build(BuildContext context) {
    screenWidth = MediaQuery.of(context).size.width;

    return Scaffold(
      backgroundColor: Color(0xFF2E2E4A),
      body: Container(
        decoration: BoxDecoration(
          image: DecorationImage(image: AssetImage(_getImage("bg_anim", format: "webp")), fit: BoxFit.fill),
        ),
        child: Column(
          children: [
            Spacer(flex: 1),
            Container(
              width: screenWidth,
              height: screenWidth,
              child: _appearsRandomlyGradientLayout(),
            ),
            Spacer(flex: 4),
          ],
        ),
      ),
    );
  }

  _appearsRandomlyGradientLayout() {
    double iconWidth = screenWidth * 110 / 375;
    return Stack(
      children: [
        Image.asset(_getImage("bg_header_ripple_anim", format: "webp"), width: double.infinity, height: double.infinity),
        Container(
          padding: EdgeInsets.symmetric(horizontal: 16), // 为了防止贴边不美观。（可以去掉）
          child: AppearsRandomlyWidget(
            width: screenWidth, // 整体的宽高。用于计算 item 出现的范围
            height: screenWidth,

            itemWidth: 56, // 每个 item 的宽高。用于计算 item 的出现
            itemHeight: 56,
            itemSizeRange: 16, // 从小变大时，最大/最小的缩放范围。（会生成随机值，此处为最大值）

            centerRestrictedArea: Size(iconWidth, iconWidth),  // 中心布局的范围
            centerChild: Container( // 中心布局。（可为空）
              width: iconWidth,
              height: iconWidth,
              alignment: Alignment.center,
              decoration: BoxDecoration(
                borderRadius: BorderRadius.circular(120),
                border: Border.all(color: Colors.white, width: 3),
              ),
              child: ClipOval(
                child: Image.asset(_getImage("header", format: "webp"), width: iconWidth, height: iconWidth),
              ),
            ),

            itemCount: list.length, // 总数
            itemBuilder: (context, index) { // item 布局
              var imgRes = list[index];
              return GestureDetector(
                child: Container(
                  decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(56),
                    border: Border.all(color: Colors.white, width: 0.5),
                  ),
                  child: ClipOval(child: Image.asset(_getImage(imgRes), width: 52, height: 52)),
                ),
                onTap: () {
                  // TODO: 点击
                },
              );
            },
          ),
        ),
      ],
    );
  }

  ///获取本地资源图片
  static String _getImage(String imageName, {String format: 'png'}) {
    return "assets/images/$imageName.$format";
  }
}
